libkbfs/tlf_edit_history.go

```diff
@@ -0,0 +1,355 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	keybase1 "github.com/keybase/client/go/protocol"
+	"golang.org/x/net/context"
+)
+
+// TlfEditNotificationType indicates what type of edit happened to a
+// file.
+type TlfEditNotificationType int
+
+const (
+	// FileCreated indicates a new file.
+	FileCreated TlfEditNotificationType = iota
+	// FileModified indicates an existing file that was written to.
+	FileModified
+	// FileDeleted indicates an existing file that was deleted.  It
+	// doesn't appear in the edit history, only in individual edit
+	// updates.
+	FileDeleted
+)
+
+// TlfEdit represents an individual update about a file edit within a
+// TLF.
+type TlfEdit struct {
+	Filepath  string // relative to the TLF root
+	Type      TlfEditNotificationType
+	LocalTime time.Time // reflects difference between server and local clock
+}
+
+const (
+	// How many edits per writer we want to return in the complete history?
+	desiredEditsPerWriter = 20
+
+	// How far back we're willing to go to get the complete history.
+	maxMDsToInspect = 1000
+)
+
+// TlfEditList is a list of edits by a particular user, that can be
+// sort by increasing timestamp.
+type TlfEditList []TlfEdit
+
+// Len implements sort.Interface for TlfEditList
+func (tel TlfEditList) Len() int {
+	return len(tel)
+}
+
+// Less implements sort.Interface for TlfEditList
+func (tel TlfEditList) Less(i, j int) bool {
+	return tel[i].LocalTime.Before(tel[j].LocalTime)
+}
+
+// Swap implements sort.Interface for TlfEditList
+func (tel TlfEditList) Swap(i, j int) {
+	tel[j], tel[i] = tel[i], tel[j]
+}
+
+// TlfWriterEdits is a map of a writer name to the most recent file
+// edits in a given folder by that writer.
+type TlfWriterEdits map[keybase1.UID]TlfEditList
+
+func (we TlfWriterEdits) isComplete() bool {
+	for _, edits := range we {
+		if len(edits) < desiredEditsPerWriter {
+			return false
+		}
+	}
+	return true
+}
+
+type writerEditEstimates map[keybase1.UID]int
+
+func (wee writerEditEstimates) isComplete() bool {
+	for _, count := range wee {
+		if count < desiredEditsPerWriter {
+			return false
+		}
+	}
+	return true
+}
+
+func (wee *writerEditEstimates) update(rmds []ImmutableRootMetadata) {
+	for i := len(rmds) - 1; i >= 0; i-- {
```

Comment:
Any particular reason for having to walk through the RMDs (I assume) backwards? If so, can you comment that? (and mention that you're assuming that the RMDs are in order of increasing revision)